In [3]:
import ROOT
import os

Welcome to JupyROOT 6.12/07


In [4]:
# !combine --help
!pwd


/uscms_data/d3/npoudyal/TTGammaSemiLeptonic13TeV/CMSSW_10_2_14/src/HiggsAnalysis/CombinedLimit/ttgammaCombineFitting/Datacard_fourTemplate_bothChannel_2016


### Important:
signal strength r = (number of signal events / number of expected signal events).    
It scales the signal up and down in such a way after considering all the nuisance parameters. Signal strength * signal = eg. expectation of number of events in the cut and count physics process.
data_obs = e.g., observed number of events in the data
POI = e.g., ttgamma crossection which is burried in signal histogram. 

### FitDiagnostics:
performs maximum likelihood fits to extract the signal yield and provide diagnostic tools such as pre and post-fit models and correlations

### MultiDimFit:
perform maximum likelihood fits in multiple parameters and likelihood scans

### Doing exercise for misIDEle

In [5]:
# myfile = ROOT.TFile("MyFiles/pho_looseCRe2e0plots_ele_2016/myMisIDEle_Prefit.root")
# myfile.ls()
# MyDataHist = myfile.Get("myData")
# MisIDEle   = myfile.Get("myMisIDEle")
# MisIDEle.SetLineWidth(2)
# MisIDEle.SetLineColor(2)
# OtherHist  = myfile.Get("myBackground")
# OtherHist.SetLineWidth(2)
# OtherHist.SetLineColor(3)

# c = ROOT.TCanvas()
# MyDataHist.Draw("hist")
# MisIDEle.Draw("same") # red
# OtherHist.Draw("same") # yellow
# c.Draw()


In [6]:
# !text2workspace.py datacard_AR.txt

In [7]:
# f_shapes_TH1 = ROOT.TFile("datacard_AR.root","READ")
# w_shapes_TH1 = f_shapes_TH1.Get("w")
# w_shapes_TH1.Print()

In [8]:
# !python ../test/printWorkspaceNormalisations.py datacard_AR.root

In [9]:
# !combine -M MultiDimFit -n AR  datacard_AR.txt --redefineSignalPOIs r,mybackground_norm

In [10]:
# !combine -M MultiDimFit -n AR  datacard_AR.txt --redefineSignalPOIs r,WGammaSF

In [11]:
!combine -M FitDiagnostics -n CR123  datacard_CR123.txt --redefineSignalPOIs r,ZGammaBkgPhotonSF,WGammaBkgPhotonSF -v2


>>> method used is FitDiagnostics
>>> random number generator seed is 123456
Creating pdfs for individual modes (2): Channel eleJets will use autoMCStats with settings: event-threshold=0, include-signal=1, hist-mode=1
Analysing bin errors for: prop_bineleJets
Poisson cut-off: 0
Bin        Contents        Error           Notes                         
0          177.953359      48.429815       total sum                     
0          14.000000       3.741657        Unweighted events, alpha=12.710954
  => Total parameter prop_bineleJets_bin0[0.00,-7.00,7.00] to be gaussian constrained
------------------------------------------------------------
1          163.547879      63.627665       total sum                     
1          7.000000        2.645751        Unweighted events, alpha=23.363983
  => Total parameter prop_bineleJets_bin1[0.00,-7.00,7.00] to be gaussian constrained
------------------------------------------------------------
2          119.233120      32.696230       total

.    1Channel muJets will use autoMCStats with settings: event-threshold=0, include-signal=1, hist-mode=1
Analysing bin errors for: prop_binmuJets
Poisson cut-off: 0
Bin        Contents        Error           Notes                         
0          217.717262      40.102846       total sum                     
0          29.000000       5.385165        Unweighted events, alpha=7.507492
  => Total parameter prop_binmuJets_bin0[0.00,-7.00,7.00] to be gaussian constrained
------------------------------------------------------------
1          285.521353      87.393760       total sum                     
1          11.000000       3.316625        Unweighted events, alpha=25.956487
  => Total parameter prop_binmuJets_bin1[0.00,-7.00,7.00] to be gaussian constrained
------------------------------------------------------------
2          165.876775      14.016027       total sum                     
2          140.000000      11.832160       Unweighted events, alpha=1.184834
  => Total par

.   done.
Importing combined pdf model_s
Missing background ModelConfig 'ModelConfig_bonly' in workspace 'w' in file roostats-O0azU4.root
Will make one from the signal ModelConfig 'ModelConfig' setting signal strenth 'r' to zero
Redefining the POIs to be: RooArgSet:: = (r,ZGammaBkgPhotonSF,WGammaBkgPhotonSF)

  RooFitResult: minimized FCN value: -23541.7, estimated distance to minimum: 0.000192356
                covariance matrix quality: Full, accurate covariance matrix
                Status : MINIMIZE=0 HESSE=0 

    Constant Parameter    Value     
  --------------------  ------------
                   ONE    1.0000e+00
  n_exp_bineleJets_proc_OtherSampleBkgPhoton    1.0000e+00
  n_exp_binmuJets_proc_OtherSampleBkgPhoton    1.0000e+00
  prop_bineleJets_bin0_In    0.0000e+00
  prop_bineleJets_bin10_In    0.0000e+00
  prop_bineleJets_bin11_In    0.0000e+00
  prop_bineleJets_bin12_In    0.0000e+00
  prop_bineleJets_bin13_In    0.0000e+00
  prop_bineleJets_bin14_In    0.0000e+00
  p

Run Minos in  Real time 0:00:00, CP time 0.080


  RooFitResult: minimized FCN value: -23541.7, estimated distance to minimum: 0.000192356
                covariance matrix quality: Full, accurate covariance matrix
                Status : MINIMIZE=0 HESSE=0 

    Constant Parameter    Value     
  --------------------  ------------
                   ONE    1.0000e+00
  n_exp_bineleJets_proc_OtherSampleBkgPhoton    1.0000e+00
  n_exp_binmuJets_proc_OtherSampleBkgPhoton    1.0000e+00
  prop_bineleJets_bin0_In    0.0000e+00
  prop_bineleJets_bin10_In    0.0000e+00
  prop_bineleJets_bin11_In    0.0000e+00
  prop_bineleJets_bin12_In    0.0000e+00
  prop_bineleJets_bin13_In    0.0000e+00
  prop_bineleJets_bin14_In    0.0000e+00
  prop_bineleJets_bin15_In    0.0000e+00
  prop_bineleJets_bin16_In    0.0000e+00
  prop_bineleJets_bin17_In    0.0000e+00
  prop_bineleJets_bin18_In    0.0000e+00
  prop_bineleJets_bin19_In    0.0000e+00
  prop_bineleJets_bin1_In    0.0000e+00
  prop_bineleJets_bin

Done in 0.01 min (cpu), 0.01 min (real)
Printing Message Summary From ... combine_logger.out
----------------------------------------------
Messages of type INFO : 9
Messages of type DEBUG : 2
Messages of type ERROR : 0
----------------------------------------------
INFO: (function: improveOnce) CascadeMinimizer.cc: 167 -- Minimisation configured with Type=Minuit2, Algo=Migrad, strategy=1, tolerance=0.1
INFO: (function: improveOnce) CascadeMinimizer.cc: 190 -- Minimisation finished with status=0
INFO: (function: improveOnce) CascadeMinimizer.cc: 195 -- Minimization success! status=0
INFO: (function: improveOnce) CascadeMinimizer.cc: 167 -- Minimisation configured with Type=Minuit2, Algo=Migrad, strategy=1, tolerance=0.1
INFO: (function: improveOnce) CascadeMinimizer.cc: 190 -- Minimisation finished with status=-1
DEBUG: (function: improveOnce) CascadeMinimizer.cc: 196 -- Minimization ended with latest status != 0 or 1
INFO: (function: improveOnce) CascadeMinimizer.cc: 167 -- Minimisati

In [12]:
!combine -M FitDiagnostics -n CR1  datacard_CR1.txt --redefineSignalPOIs r,ZGammaBkgPhotonSF,WGammaBkgPhotonSF -v2


>>> method used is FitDiagnostics
>>> random number generator seed is 123456
Creating pdfs for individual modes (2): Channel eleJets will use autoMCStats with settings: event-threshold=0, include-signal=1, hist-mode=1
Analysing bin errors for: prop_bineleJets
Poisson cut-off: 0
Bin        Contents        Error           Notes                         
0          144.168633      47.848889       total sum                     
0          9.000000        3.000000        Unweighted events, alpha=16.018737
  => Total parameter prop_bineleJets_bin0[0.00,-7.00,7.00] to be gaussian constrained
------------------------------------------------------------
1          133.413374      63.252003       total sum                     
1          4.000000        2.000000        Unweighted events, alpha=33.353343
  => Total parameter prop_bineleJets_bin1[0.00,-7.00,7.00] to be gaussian constrained
------------------------------------------------------------
2          81.382341       31.625106       total

.    1Channel muJets will use autoMCStats with settings: event-threshold=0, include-signal=1, hist-mode=1
Analysing bin errors for: prop_binmuJets
Poisson cut-off: 0
Bin        Contents        Error           Notes                         
0          155.061186      39.387895       total sum                     
0          15.000000       3.872983        Unweighted events, alpha=10.337412
  => Total parameter prop_binmuJets_bin0[0.00,-7.00,7.00] to be gaussian constrained
------------------------------------------------------------
1          211.237432      85.280096       total sum                     
1          6.000000        2.449490        Unweighted events, alpha=35.206239
  => Total parameter prop_binmuJets_bin1[0.00,-7.00,7.00] to be gaussian constrained
------------------------------------------------------------
2          97.862687       11.143438       total sum                     
2          77.000000       8.774964        Unweighted events, alpha=1.270944
  => Total pa

.   done.
Importing combined pdf model_s
Missing background ModelConfig 'ModelConfig_bonly' in workspace 'w' in file roostats-owvKzU.root
Will make one from the signal ModelConfig 'ModelConfig' setting signal strenth 'r' to zero
Redefining the POIs to be: RooArgSet:: = (r,ZGammaBkgPhotonSF,WGammaBkgPhotonSF)

  RooFitResult: minimized FCN value: -19868.1, estimated distance to minimum: 0.000145183
                covariance matrix quality: Full, accurate covariance matrix
                Status : MINIMIZE=0 HESSE=0 

    Constant Parameter    Value     
  --------------------  ------------
                   ONE    1.0000e+00
  n_exp_bineleJets_proc_OtherSampleBkgPhoton    1.0000e+00
  n_exp_binmuJets_proc_OtherSampleBkgPhoton    1.0000e+00
  prop_bineleJets_bin0_In    0.0000e+00
  prop_bineleJets_bin10_In    0.0000e+00
  prop_bineleJets_bin11_In    0.0000e+00
  prop_bineleJets_bin12_In    0.0000e+00
  prop_bineleJets_bin13_In    0.0000e+00
  prop_bineleJets_bin14_In    0.0000e+00
  p

Run Minos in  Real time 0:00:00, CP time 0.040


  RooFitResult: minimized FCN value: -19868.1, estimated distance to minimum: 0.000145183
                covariance matrix quality: Full, accurate covariance matrix
                Status : MINIMIZE=0 HESSE=0 

    Constant Parameter    Value     
  --------------------  ------------
                   ONE    1.0000e+00
  n_exp_bineleJets_proc_OtherSampleBkgPhoton    1.0000e+00
  n_exp_binmuJets_proc_OtherSampleBkgPhoton    1.0000e+00
  prop_bineleJets_bin0_In    0.0000e+00
  prop_bineleJets_bin10_In    0.0000e+00
  prop_bineleJets_bin11_In    0.0000e+00
  prop_bineleJets_bin12_In    0.0000e+00
  prop_bineleJets_bin13_In    0.0000e+00
  prop_bineleJets_bin14_In    0.0000e+00
  prop_bineleJets_bin15_In    0.0000e+00
  prop_bineleJets_bin16_In    0.0000e+00
  prop_bineleJets_bin17_In    0.0000e+00
  prop_bineleJets_bin18_In    0.0000e+00
  prop_bineleJets_bin19_In    0.0000e+00
  prop_bineleJets_bin1_In    0.0000e+00
  prop_bineleJets_bin

Done in 0.01 min (cpu), 0.02 min (real)
Printing Message Summary From ... combine_logger.out
----------------------------------------------
Messages of type INFO : 9
Messages of type DEBUG : 2
Messages of type ERROR : 0
----------------------------------------------
INFO: (function: improveOnce) CascadeMinimizer.cc: 167 -- Minimisation configured with Type=Minuit2, Algo=Migrad, strategy=1, tolerance=0.1
INFO: (function: improveOnce) CascadeMinimizer.cc: 190 -- Minimisation finished with status=0
INFO: (function: improveOnce) CascadeMinimizer.cc: 195 -- Minimization success! status=0
INFO: (function: improveOnce) CascadeMinimizer.cc: 167 -- Minimisation configured with Type=Minuit2, Algo=Migrad, strategy=1, tolerance=0.1
INFO: (function: improveOnce) CascadeMinimizer.cc: 190 -- Minimisation finished with status=-1
DEBUG: (function: improveOnce) CascadeMinimizer.cc: 196 -- Minimization ended with latest status != 0 or 1
INFO: (function: improveOnce) CascadeMinimizer.cc: 167 -- Minimisati

In [13]:
!combine -M FitDiagnostics -n CR2  datacard_CR2.txt --redefineSignalPOIs r,ZGammaBkgPhotonSF,WGammaBkgPhotonSF -v2


>>> method used is FitDiagnostics
>>> random number generator seed is 123456
Creating pdfs for individual modes (2): Channel eleJets will use autoMCStats with settings: event-threshold=0, include-signal=1, hist-mode=1
Analysing bin errors for: prop_bineleJets
Poisson cut-off: 0
Bin        Contents        Error           Notes                         
0          21.853806       7.181898        total sum                     
0          9.000000        3.000000        Unweighted events, alpha=2.428201
  => Total parameter prop_bineleJets_bin0[0.00,-7.00,7.00] to be gaussian constrained
------------------------------------------------------------
1          14.870016       5.134427        total sum                     
1          8.000000        2.828427        Unweighted events, alpha=1.858752
  => Total parameter prop_bineleJets_bin1[0.00,-7.00,7.00] to be gaussian constrained
------------------------------------------------------------
2          25.175880       8.049666        total s

.    1Channel muJets will use autoMCStats with settings: event-threshold=0, include-signal=1, hist-mode=1
Analysing bin errors for: prop_binmuJets
Poisson cut-off: 0
Bin        Contents        Error           Notes                         
0          39.513010       5.993179        total sum                     
0          43.000000       6.557439        Unweighted events, alpha=0.918907
  => Total parameter prop_binmuJets_bin0[0.00,-7.00,7.00] to be gaussian constrained
------------------------------------------------------------
1          56.025411       18.970050       total sum                     
1          9.000000        3.000000        Unweighted events, alpha=6.225046
  => Total parameter prop_binmuJets_bin1[0.00,-7.00,7.00] to be gaussian constrained
------------------------------------------------------------
2          40.882767       5.882400        total sum                     
2          48.000000       6.928203        Unweighted events, alpha=0.851724
  => Total para

.   done.
Importing combined pdf model_s
Missing background ModelConfig 'ModelConfig_bonly' in workspace 'w' in file roostats-naCbZw.root
Will make one from the signal ModelConfig 'ModelConfig' setting signal strenth 'r' to zero
Redefining the POIs to be: RooArgSet:: = (r,ZGammaBkgPhotonSF,WGammaBkgPhotonSF)

  RooFitResult: minimized FCN value: -7545.05, estimated distance to minimum: 5.26003e-05
                covariance matrix quality: Full, accurate covariance matrix
                Status : MINIMIZE=0 HESSE=0 

    Constant Parameter    Value     
  --------------------  ------------
                   ONE    1.0000e+00
  n_exp_bineleJets_proc_OtherSampleBkgPhoton    1.0000e+00
  n_exp_binmuJets_proc_OtherSampleBkgPhoton    1.0000e+00
  prop_bineleJets_bin0_In    0.0000e+00
  prop_bineleJets_bin10_In    0.0000e+00
  prop_bineleJets_bin11_In    0.0000e+00
  prop_bineleJets_bin12_In    0.0000e+00
  prop_bineleJets_bin13_In    0.0000e+00
  prop_bineleJets_bin14_In    0.0000e+00
  p


  RooFitResult: minimized FCN value: -4400.34, estimated distance to minimum: 3.16967e-05
                covariance matrix quality: Full, accurate covariance matrix
                Status : MINIMIZE=0 HESSE=0 

    Constant Parameter    Value     
  --------------------  ------------
                   ONE    1.0000e+00
  n_exp_bineleJets_proc_OtherSampleBkgPhoton    1.0000e+00
  n_exp_binmuJets_proc_OtherSampleBkgPhoton    1.0000e+00
  prop_bineleJets_bin0_In    0.0000e+00
  prop_bineleJets_bin10_In    0.0000e+00
  prop_bineleJets_bin11_In    0.0000e+00
  prop_bineleJets_bin12_In    0.0000e+00
  prop_bineleJets_bin13_In    0.0000e+00
  prop_bineleJets_bin14_In    0.0000e+00
  prop_bineleJets_bin15_In    0.0000e+00
  prop_bineleJets_bin16_In    0.0000e+00
  prop_bineleJets_bin17_In    0.0000e+00
  prop_bineleJets_bin18_In    0.0000e+00
  prop_bineleJets_bin19_In    0.0000e+00
  prop_bineleJets_bin1_In    0.0000e+00
  prop_bineleJets_bin20_In    0.0000e+00
  prop_bineleJets_bin21_In  

Run Minos in  Real time 0:00:00, CP time 0.080


  RooFitResult: minimized FCN value: -4400.34, estimated distance to minimum: 3.16967e-05
                covariance matrix quality: Full, accurate covariance matrix
                Status : MINIMIZE=0 HESSE=0 

    Constant Parameter    Value     
  --------------------  ------------
                   ONE    1.0000e+00
  n_exp_bineleJets_proc_OtherSampleBkgPhoton    1.0000e+00
  n_exp_binmuJets_proc_OtherSampleBkgPhoton    1.0000e+00
  prop_bineleJets_bin0_In    0.0000e+00
  prop_bineleJets_bin10_In    0.0000e+00
  prop_bineleJets_bin11_In    0.0000e+00
  prop_bineleJets_bin12_In    0.0000e+00
  prop_bineleJets_bin13_In    0.0000e+00
  prop_bineleJets_bin14_In    0.0000e+00
  prop_bineleJets_bin15_In    0.0000e+00
  prop_bineleJets_bin16_In    0.0000e+00
  prop_bineleJets_bin17_In    0.0000e+00
  prop_bineleJets_bin18_In    0.0000e+00
  prop_bineleJets_bin19_In    0.0000e+00
  prop_bineleJets_bin1_In    0.0000e+00
  prop_bineleJets_bin

Done in 0.01 min (cpu), 0.01 min (real)
Printing Message Summary From ... combine_logger.out
----------------------------------------------
Messages of type INFO : 10
Messages of type DEBUG : 2
Messages of type ERROR : 0
----------------------------------------------
INFO: (function: improveOnce) CascadeMinimizer.cc: 167 -- Minimisation configured with Type=Minuit2, Algo=Migrad, strategy=1, tolerance=0.1
INFO: (function: improveOnce) CascadeMinimizer.cc: 190 -- Minimisation finished with status=0
INFO: (function: improveOnce) CascadeMinimizer.cc: 195 -- Minimization success! status=0
INFO: (function: improveOnce) CascadeMinimizer.cc: 167 -- Minimisation configured with Type=Minuit2, Algo=Migrad, strategy=1, tolerance=0.1
INFO: (function: improveOnce) CascadeMinimizer.cc: 190 -- Minimisation finished with status=0
INFO: (function: improveOnce) CascadeMinimizer.cc: 195 -- Minimization success! status=0
DEBUG: (function: runSpecific) FitDiagnostics.cc: 304 -- Fit B-only, status = 0, numBa

In [14]:
!combine -M FitDiagnostics -n CR3  datacard_CR3.txt --redefineSignalPOIs r,ZGammaBkgPhotonSF,WGammaBkgPhotonSF -v2


>>> method used is FitDiagnostics
>>> random number generator seed is 123456
Creating pdfs for individual modes (2): Channel eleJets will use autoMCStats with settings: event-threshold=0, include-signal=1, hist-mode=1
Analysing bin errors for: prop_bineleJets
Poisson cut-off: 0
Bin        Contents        Error           Notes                         
0          11.930910       2.085956        total sum                     
0          33.000000       5.744563        Unweighted events, alpha=0.361543
  => Total parameter prop_bineleJets_bin0[0.00,-7.00,7.00] to be gaussian constrained
------------------------------------------------------------
1          15.264498       4.615367        total sum                     
1          11.000000       3.316625        Unweighted events, alpha=1.387682
  => Total parameter prop_bineleJets_bin1[0.00,-7.00,7.00] to be gaussian constrained
------------------------------------------------------------
2          12.674878       2.024597        total s

.    1Channel muJets will use autoMCStats with settings: event-threshold=0, include-signal=1, hist-mode=1
Analysing bin errors for: prop_binmuJets
Poisson cut-off: 0
Bin        Contents        Error           Notes                         
0          23.143089       4.573161        total sum                     
0          26.000000       5.099020        Unweighted events, alpha=0.890119
  => Total parameter prop_binmuJets_bin0[0.00,-7.00,7.00] to be gaussian constrained
------------------------------------------------------------
1          18.258527       2.260904        total sum                     
1          65.000000       8.062258        Unweighted events, alpha=0.280900
  => Total parameter prop_binmuJets_bin1[0.00,-7.00,7.00] to be gaussian constrained
------------------------------------------------------------
2          27.131302       6.137602        total sum                     
2          20.000000       4.472136        Unweighted events, alpha=1.356565
  => Total para

.   done.
Importing combined pdf model_s
Missing background ModelConfig 'ModelConfig_bonly' in workspace 'w' in file roostats-KdGr3s.root
Will make one from the signal ModelConfig 'ModelConfig' setting signal strenth 'r' to zero
Redefining the POIs to be: RooArgSet:: = (r,ZGammaBkgPhotonSF,WGammaBkgPhotonSF)

  RooFitResult: minimized FCN value: -2164.7, estimated distance to minimum: 6.79033e-06
                covariance matrix quality: Full, accurate covariance matrix
                Status : MINIMIZE=0 HESSE=0 

    Constant Parameter    Value     
  --------------------  ------------
                   ONE    1.0000e+00
  n_exp_bineleJets_proc_OtherSampleBkgPhoton    1.0000e+00
  n_exp_binmuJets_proc_OtherSampleBkgPhoton    1.0000e+00
  prop_bineleJets_bin0_In    0.0000e+00
  prop_bineleJets_bin10_In    0.0000e+00
  prop_bineleJets_bin11_In    0.0000e+00
  prop_bineleJets_bin12_In    0.0000e+00
  prop_bineleJets_bin13_In    0.0000e+00
  prop_bineleJets_bin14_In    0.0000e+00
  pr


  RooFitResult: minimized FCN value: -1055.58, estimated distance to minimum: 2.65123e-05
                covariance matrix quality: Full, accurate covariance matrix
                Status : MINIMIZE=0 HESSE=0 

    Constant Parameter    Value     
  --------------------  ------------
                   ONE    1.0000e+00
  n_exp_bineleJets_proc_OtherSampleBkgPhoton    1.0000e+00
  n_exp_binmuJets_proc_OtherSampleBkgPhoton    1.0000e+00
  prop_bineleJets_bin0_In    0.0000e+00
  prop_bineleJets_bin10_In    0.0000e+00
  prop_bineleJets_bin11_In    0.0000e+00
  prop_bineleJets_bin12_In    0.0000e+00
  prop_bineleJets_bin13_In    0.0000e+00
  prop_bineleJets_bin14_In    0.0000e+00
  prop_bineleJets_bin15_In    0.0000e+00
  prop_bineleJets_bin16_In    0.0000e+00
  prop_bineleJets_bin17_In    0.0000e+00
  prop_bineleJets_bin18_In    0.0000e+00
  prop_bineleJets_bin19_In    0.0000e+00
  prop_bineleJets_bin1_In    0.0000e+00
  prop_bineleJets_bin20_In    0.0000e+00
  prop_bineleJets_bin21_In  

Run Minos in  Real time 0:00:00, CP time 0.050


  RooFitResult: minimized FCN value: -1055.58, estimated distance to minimum: 2.65123e-05
                covariance matrix quality: Full, accurate covariance matrix
                Status : MINIMIZE=0 HESSE=0 

    Constant Parameter    Value     
  --------------------  ------------
                   ONE    1.0000e+00
  n_exp_bineleJets_proc_OtherSampleBkgPhoton    1.0000e+00
  n_exp_binmuJets_proc_OtherSampleBkgPhoton    1.0000e+00
  prop_bineleJets_bin0_In    0.0000e+00
  prop_bineleJets_bin10_In    0.0000e+00
  prop_bineleJets_bin11_In    0.0000e+00
  prop_bineleJets_bin12_In    0.0000e+00
  prop_bineleJets_bin13_In    0.0000e+00
  prop_bineleJets_bin14_In    0.0000e+00
  prop_bineleJets_bin15_In    0.0000e+00
  prop_bineleJets_bin16_In    0.0000e+00
  prop_bineleJets_bin17_In    0.0000e+00
  prop_bineleJets_bin18_In    0.0000e+00
  prop_bineleJets_bin19_In    0.0000e+00
  prop_bineleJets_bin1_In    0.0000e+00
  prop_bineleJets_bin

Done in 0.01 min (cpu), 0.01 min (real)
Printing Message Summary From ... combine_logger.out
----------------------------------------------
Messages of type INFO : 10
Messages of type DEBUG : 2
Messages of type ERROR : 0
----------------------------------------------
INFO: (function: improveOnce) CascadeMinimizer.cc: 167 -- Minimisation configured with Type=Minuit2, Algo=Migrad, strategy=1, tolerance=0.1
INFO: (function: improveOnce) CascadeMinimizer.cc: 190 -- Minimisation finished with status=0
INFO: (function: improveOnce) CascadeMinimizer.cc: 195 -- Minimization success! status=0
INFO: (function: improveOnce) CascadeMinimizer.cc: 167 -- Minimisation configured with Type=Minuit2, Algo=Migrad, strategy=1, tolerance=0.1
INFO: (function: improveOnce) CascadeMinimizer.cc: 190 -- Minimisation finished with status=0
INFO: (function: improveOnce) CascadeMinimizer.cc: 195 -- Minimization success! status=0
DEBUG: (function: runSpecific) FitDiagnostics.cc: 304 -- Fit B-only, status = 0, numBa

In [15]:
!combine -M FitDiagnostics -n CR4  datacard_CR4.txt --redefineSignalPOIs r,ZGammaBkgPhotonSF,WGammaBkgPhotonSF -v2


>>> method used is FitDiagnostics
>>> random number generator seed is 123456
Creating pdfs for individual modes (2): Channel eleJets will use autoMCStats with settings: event-threshold=0, include-signal=1, hist-mode=1
Analysing bin errors for: prop_bineleJets
Poisson cut-off: 0
Bin        Contents        Error           Notes                         
0          34.573583       8.101353        total sum                     
0          18.000000       4.242641        Unweighted events, alpha=1.920755
  => Total parameter prop_bineleJets_bin0[0.00,-7.00,7.00] to be gaussian constrained
------------------------------------------------------------
1          22.325551       7.881637        total sum                     
1          8.000000        2.828427        Unweighted events, alpha=2.790694
  => Total parameter prop_bineleJets_bin1[0.00,-7.00,7.00] to be gaussian constrained
------------------------------------------------------------
2          24.777635       7.189154        total s

.    1Channel muJets will use autoMCStats with settings: event-threshold=0, include-signal=1, hist-mode=1
Analysing bin errors for: prop_binmuJets
Poisson cut-off: 0
Bin        Contents        Error           Notes                         
0          47.199055       6.672904        total sum                     
0          50.000000       7.071068        Unweighted events, alpha=0.943981
  => Total parameter prop_binmuJets_bin0[0.00,-7.00,7.00] to be gaussian constrained
------------------------------------------------------------
1          66.023922       14.313866       total sum                     
1          21.000000       4.582576        Unweighted events, alpha=3.143996
  => Total parameter prop_binmuJets_bin1[0.00,-7.00,7.00] to be gaussian constrained
------------------------------------------------------------
2          66.779813       9.704524        total sum                     
2          47.000000       6.855655        Unweighted events, alpha=1.420847
  => Total para

.   done.
Importing combined pdf model_s
Missing background ModelConfig 'ModelConfig_bonly' in workspace 'w' in file roostats-8eueIs.root
Will make one from the signal ModelConfig 'ModelConfig' setting signal strenth 'r' to zero
Redefining the POIs to be: RooArgSet:: = (r,ZGammaBkgPhotonSF,WGammaBkgPhotonSF)

  RooFitResult: minimized FCN value: -3969.44, estimated distance to minimum: 7.74071e-05
                covariance matrix quality: Full, accurate covariance matrix
                Status : MINIMIZE=0 HESSE=0 

    Constant Parameter    Value     
  --------------------  ------------
                   ONE    1.0000e+00
  n_exp_bineleJets_proc_OtherSampleBkgPhoton    1.0000e+00
  n_exp_binmuJets_proc_OtherSampleBkgPhoton    1.0000e+00
  prop_bineleJets_bin0_In    0.0000e+00
  prop_bineleJets_bin10_In    0.0000e+00
  prop_bineleJets_bin11_In    0.0000e+00
  prop_bineleJets_bin12_In    0.0000e+00
  prop_bineleJets_bin13_In    0.0000e+00
  prop_bineleJets_bin14_In    0.0000e+00
  p


 --- FitDiagnostics ---
Fit failed.
Done in 0.01 min (cpu), 0.01 min (real)
Printing Message Summary From ... combine_logger.out
----------------------------------------------
Messages of type INFO : 8
Messages of type DEBUG : 2
Messages of type ERROR : 0
----------------------------------------------
INFO: (function: improveOnce) CascadeMinimizer.cc: 167 -- Minimisation configured with Type=Minuit2, Algo=Migrad, strategy=1, tolerance=0.1
INFO: (function: improveOnce) CascadeMinimizer.cc: 190 -- Minimisation finished with status=0
INFO: (function: improveOnce) CascadeMinimizer.cc: 195 -- Minimization success! status=0
INFO: (function: improveOnce) CascadeMinimizer.cc: 167 -- Minimisation configured with Type=Minuit2, Algo=Migrad, strategy=1, tolerance=0.1
INFO: (function: improveOnce) CascadeMinimizer.cc: 190 -- Minimisation finished with status=0
INFO: (function: improveOnce) CascadeMinimizer.cc: 195 -- Minimization success! status=0
DEBUG: (function: runSpecific) FitDiagnostics.cc: 

In [16]:
# !combine -M MultiDimFit -n CR5  datacard_CR5.txt --redefineSignalPOIs r,WGammaSF

In [17]:
# !combine -M MultiDimFit -n CR6  datacard_CR6.txt --redefineSignalPOIs r,WGammaSF

In [18]:
# !combine -M MultiDimFit -n CR7  datacard_CR7.txt --redefineSignalPOIs r,WGammaSF

In [19]:
# !combine -M MultiDimFit -n SR8  datacard_SR8.txt --redefineSignalPOIs r,WGammaSF

In [20]:
# f_input_shapes_TH1 = ROOT.TFile("simple-shapes-TH1_input.root","READ")
# f_input_shapes_TH1.ls()


In [21]:
# h_background = f_input_shapes_TH1.Get("background")
# h_background.SetLineColor(ROOT.kBlack)
# print h_background.Integral(-1,-1)

# h_background_alphaUp = f_input_shapes_TH1.Get("background_alphaUp")
# h_background_alphaUp.SetLineColor(ROOT.kBlue)
# print h_background_alphaUp.Integral(-1,-1)

# h_background_alphaDown = f_input_shapes_TH1.Get("background_alphaDown")
# h_background_alphaDown.SetLineColor(ROOT.kRed)
# print h_background_alphaDown.Integral(-1,-1)

# c = ROOT.TCanvas()
# h_background_alphaDown.Draw()
# h_background.Draw("same")
# h_background_alphaUp.Draw("same")

# c.Draw()

In [22]:
# h_signal = f_input_shapes_TH1.Get("signal")
# h_signal.SetLineColor(ROOT.kBlack)
# print h_signal.Integral()

# h_signal_sigmaUp = f_input_shapes_TH1.Get("signal_sigmaUp")
# h_signal_sigmaUp.SetLineColor(ROOT.kBlue)
# print h_signal_sigmaUp.Integral()

# h_signal_sigmaDown = f_input_shapes_TH1.Get("signal_sigmaDown")
# h_signal_sigmaDown.SetLineColor(ROOT.kRed)
# print h_signal_sigmaDown.Integral()

# c = ROOT.TCanvas()
# h_signal_sigmaDown.Draw()
# h_signal.Draw("same")
# h_signal_sigmaUp.Draw("same")

# c.Draw()

The combine tool converts datacards into a RooWorkspace to build the likelihood

In [23]:
# !text2workspace.py ttgamma_sys.txt

In [24]:
# f_shapes_TH1 = ROOT.TFile("ttgamma_sys.root","READ")
# w_shapes_TH1 = f_shapes_TH1.Get("w")
# w_shapes_TH1.Print()

The workspace contains "model_s" i.e. the likelihood function, which is composed of the observable "CMS_th1x" which is split in several channels, the POI "r" and several nuisance parameters. Lets see how combine smoothly morphs the shape of the background as a function of the nuisance parameter alpha:


In [25]:
# shapeBkg = w_shapes_TH1.pdf("shapeBkg_mybin_background_morph")
# th1x = w_shapes_TH1.var("CMS_th1x")
# plot_th1x = th1x.frame()
# alpha = w_shapes_TH1.var("alpha")
# alpha.setVal(0.0)
# shapeBkg.plotOn(plot_th1x,ROOT.RooFit.LineColor(ROOT.kBlack))
# alpha.setVal(0.5)
# shapeBkg.plotOn(plot_th1x,ROOT.RooFit.LineColor(ROOT.kGreen))
# alpha.setVal(1.0)
# shapeBkg.plotOn(plot_th1x,ROOT.RooFit.LineColor(ROOT.kRed))
# alpha.setVal(2.0)
# shapeBkg.plotOn(plot_th1x,ROOT.RooFit.LineColor(ROOT.kOrange))

# c2 = ROOT.TCanvas()
# plot_th1x.Draw()
# c2.Draw()

In [26]:
# shapeSig = w_shapes_TH1.pdf("shapeSig_mybin_signal_morph")
# th1x = w_shapes_TH1.var("CMS_th1x")
# plot_th1x = th1x.frame()
# sigma = w_shapes_TH1.var("sigma")
# sigma.setVal(0.0)
# shapeSig.plotOn(plot_th1x,ROOT.RooFit.LineColor(ROOT.kBlack))
# sigma.setVal(0.5)
# shapeSig.plotOn(plot_th1x,ROOT.RooFit.LineColor(ROOT.kGreen))
# sigma.setVal(1.0)
# shapeSig.plotOn(plot_th1x,ROOT.RooFit.LineColor(ROOT.kRed))
# sigma.setVal(2.0)
# shapeSig.plotOn(plot_th1x,ROOT.RooFit.LineColor(ROOT.kOrange))

# c2 = ROOT.TCanvas()
# plot_th1x.Draw()
# c2.Draw()

Prints the workspace normalization 

In [27]:
# !python ../test/printWorkspaceNormalisations.py ttgamma_sys.root

If we redefine POI then it will be float and can take any values. If we have things in datacard then everything will be used to produce the best r value.  
The option "--run both" tells combine to run both expected and observed limits.You can see we have computed an observed limit on r, and also the nominal expected, and +/- 1 and 2 sigma expected limits. [-s -1 == random seed]

In [28]:
# !combine -M MultiDimFit -n Nabin ttgamma_sys.txt

In [29]:
# !combine -n LimitTestNabin -M AsymptoticLimits ttgamma_sys.root --run both

In [30]:
# asymptoticfile = ROOT.TFile("higgsCombineLimitTestNabin.AsymptoticLimits.mH120.root")
# limit = asymptoticfile.Get("limit")
# for i in xrange(limit.GetEntries()):
#     limit.GetEntry(i)
#     print i,limit.mh,limit.limit

1-sigma ==  68%, 2-sigma == 95%, 3-sigma == 99.7%. The higher sigma value means that the discovery is less likely to be accidently a mistake or 'random chance'.   
How do we test hypothesis?  
Null hypothesis  means A and B are not connected. Research hypothesis means A and B are connected. First assume Null hypothesis is true. Then calculate the probability of obtaining data as extreme as the kind they are observing. 
P-value close to zero means Null hypothesis is false, however, larger p-value means Null hypothesis is true. For example, p-value = 5% means that chance that the data sample being tested is drawn from where Null hypothesis is true.  
Next example, the chances of a result being greater than 1-sigma is 15.9%, 2-sigma is 2.3%, 3-sigma is 0.2%. 